homeassistant/components/enasolar/sensor.py

```diff
@@ -0,0 +1,250 @@
+            async_track_time_interval_backoff(
+                hass, async_enasolar_meters, SCAN_METERS_MIN_INTERVAL
+            )
+        )
+        config_entry.async_on_unload(
+            async_track_time_interval_backoff(
+                hass, async_enasolar_data, SCAN_DATA_MIN_INTERVAL
+            )
+        )
+
+    if hass.is_running:
+        start_update_interval(None)
+    else:
+        hass.bus.async_listen_once(EVENT_HOMEASSISTANT_START, start_update_interval)
+
+
+@callback
+def async_track_time_interval_backoff(hass, action, min_interval) -> CALLBACK_TYPE:
+    """Add a listener that fires repetitively and increases the interval when failed."""
+    remove = None
+    interval = min_interval
+
+    async def interval_listener(now=None):  # pylint: disable=unused-argument
+        """Handle elapsed interval with backoff."""
+        nonlocal interval, remove
+
+        try:
+            if await action():
+                interval = min_interval
+            else:
+                interval = min(interval * 2, SCAN_MAX_INTERVAL)
+        finally:
+            remove = async_call_later(hass, interval, interval_listener)
+
+    hass.async_create_task(interval_listener())
+
+    def remove_listener():
+        """Remove interval listener."""
+        remove()  # pylint: disable=not-callable
+
+    return remove_listener
+
+
+class EnaSolarSensor(SensorEntity):
+    """Representation of a EnaSolar sensor."""
+
+    def __init__(self, pyenasolar_sensor, inverter_name=None, serial_no=None):
+        """Initialize the EnaSolar sensor."""
+        self.sensor = pyenasolar_sensor
+        if inverter_name:
+            self._attr_name = f"{inverter_name}_{self.sensor.name}"
+        else:
+            self._attr_name = f"enasolar_{self.sensor.name}"
+        self.serial_no = serial_no
+        self._native_value = self.sensor.value
+
+        if pyenasolar_sensor.is_meter:
+            self._attr_state_class = STATE_CLASS_MEASUREMENT
+        else:
+            self._attr_state_class = STATE_CLASS_TOTAL_INCREASING
+        self._attr_native_unit_of_measurement = ENASOLAR_UNIT_MAPPINGS[self.sensor.unit]
+        self._attr_should_poll = False
+
+    @property
+    def name(self) -> str | None:
+        """Return the name of the sensor."""
+        if hasattr(self, "_attr_name"):
+            return self._attr_name
+        return None
+
+    @property
+    def native_value(self):
+        """Return the current sensor value."""
+        return self._native_value
+
+    @property
+    def device_class(self):
+        """Return the device class the sensor belongs to."""
+        if self._attr_native_unit_of_measurement == POWER_KILO_WATT:
+            return DEVICE_CLASS_POWER
+        if self._attr_native_unit_of_measurement == ENERGY_KILO_WATT_HOUR:
+            return DEVICE_CLASS_ENERGY
+        if self._attr_unit_of_measurement in (TEMP_CELSIUS, TEMP_FAHRENHEIT):
+            return DEVICE_CLASS_TEMPERATURE
+        return None
```

Comment:
Things like these makes me wonder if you should implement `SensorEntityDescription`s instead?

Reply:
I started this integration by basing it on an exisiting one 'saj' as it seemed to do the sorts of things needed.  It obviously was not a good choice as it doesn't conform with the coding guidelines that now are required. I freely admit to not understanding the internals of HA and how the various things interact, my goal was to get something up and running I could use and then I decided that it would possibly be useful for other people as well as EnaSolar was a popular inverter brand in the Southern Hemisphere.